---
title: "INVAR2 Analysis"
output:
  html_document:
    dev: 'svg'
    df_print: 'tibble'
    toc: TRUE
    toc_float: TRUE
---

INVAR2 aims to detect and quantify known patient-specific mutations in custom capture liquid biopsy data. 

Pileups are performed with the input reads at patient-specific loci, which are annotated with trinucleotide context-specific error rates, 1000 genomes and COSMIC SNP information, and tumour allele fraction. The annotated dataframe is then split between on target data for patient-specific locations and non-patient-specific data based on initial mutation list. The size information of each of the mutant and wild-type reads is determined for the future weighing of mutations. A patient specific outlier suppression calculation is applied, which identifies noise based on knowledge of a large number of patient-specific loci. The INVAR2 algorithm then generates a score for each sample which includes the features mentioned above. Control samples are used to determine a threshold for the INVAR score.

This document summarizes and explains the results of the pipeline.


## Input data overview

## Mutations per patient captured and passing all pipeline filters

```{r fig.width = 6, fig.height = 7, echo = FALSE}
plots$P3
```


## On Target Locus Error Rates

Locus error rate = overall background error rate per locus, aggregated across control samples

```{r fig.width = 11, fig.height = 7, echo = FALSE}
plots$P0
```

## Cohort by Mutant Context

```{r fig.width = 6, fig.height = 5, echo = FALSE}
plots$P1
```

## AF by mutation class

```{r fig.width = 6, fig.height = 7, echo = FALSE}
plots$P2
```


## Mutation class distribution by cohort

```{r fig.width = 6, fig.height = 5, echo = FALSE}
plots$P4
```

## Case vs control error rates

```{r fig.width = 8, fig.height = 4, echo = FALSE}
plots$P5
```

# Summary Plots

## Background error rates

```{r fig.width = 8, fig.height = 4, echo = FALSE}
plots$P7
```

## Error rate comparison - both strands

```{r fig.width = 6, fig.height = 3, echo = FALSE}
plots$P9a
```

## Error rate comparison - locus noise

```{r fig.width = 6, fig.height = 3, echo = FALSE}
plots$P9b
```

## Error rate comparison - locus noise &amp; both strands

```{r fig.width = 6, fig.height = 3, echo = FALSE}
plots$P9c
```

## Filtering comparison

```{r fig.width = 6, fig.height = 4, echo = FALSE}
plots$P20
```

# Outlier suppression plots

## Data retained

```{r fig.width = 5, fig.height = 6, echo = FALSE}
plots$P6
```

## Background polishing

```{r fig.width = 15, fig.height = 12, echo = FALSE}
plots$P8
```

# Tumour AF in observed and unobserved loci.

```{r fig.width = 5, fig.height = 4, echo = FALSE}
plots$P10
```

# Fragment size per cohort

With different levels of error-suppression.

```{r fig.width = 6, fig.height = 5, echo = FALSE}
plots$P11
```

# Enrichment level

```{r fig.width = 6, fig.height = 5, echo = FALSE}
if (is.null(plots$P12)) {
    print("There are no data points from which to create the enrichment level plot.")
} else {
    plots$P12
}
```

# Receiver Operating Characteristic (ROC) Plots

With size:

```{r fig.width = 4, fig.height = 3, echo = FALSE}
plots$P13a
```

Without size:

```{r fig.width = 4, fig.height = 3, echo = FALSE}
plots$P13b
```

# IR (depth) to IMAF

```{r fig.width = 6, fig.height = 4, echo = FALSE}
if (is.null(plots$P14)) {
    warning("There is no IF patient data from which to create the IF to IMAF plot.")
} else {
    plots$P14
}
```

# Waterfall Plots

## Detectable vs non detectable using dPCR

```{r fig.width = 10, fig.height = 5, echo = FALSE}
if (is.null(plots$P15)) {
    warning("There is no patient specific GLRT data from which to create the detectable/non detectable plot.")
} else {
    plots$P15
}
```

## Cancer genomes

```{r fig.width = 10, fig.height = 5, echo = FALSE}
if (is.null(plots$P16)) {
    warning("There is no patient specific GLRT data from which to create the cancer genomes waterfall plot.")
} else {
    plots$P16
}
```

# dPCR comparison plot

```{r fig.width = 8, fig.height = 5, echo = FALSE}
if (is.null(plots$P17)) {
    warning("There is no patient specific GLRT data from which to create the dPCR comparison plot.")
} else {
    plots$P17
}
```

# Mutation Tracking

```{r, echo = FALSE}
if (nrow(mutationTrackingTable) > 0) {
    toYesNo <- function(v) { ifelse(v, 'Yes', 'No' ) }
    mutationTrackingTable %>%
        mutate_if(is.logical, toYesNo) %>%
        kable()
}
```
